include/vkgcDefs.h

```diff
@@ -749,4 +761,208 @@ class IPipelineDumper {
+  virtual Result SetValue(void *rawHandle, bool success, const void *pData, size_t dataLen) = 0;
+
+  /// \brief Populate the value contents of a cache entry and release the handle.
+  ///
+  /// Semantics are identical to SetValue, except that the handle is guaranteed to be released.
+  /// Doing this atomically can sometimes allow a more efficient implementation; the default
+  /// implementation is trivial.
+  virtual Result PutWithValue(void *rawHandle, bool success, const void* pData, size_t dataLen)
+  {
+    Result result = Result::ErrorUnknown;
+    if (rawHandle == nullptr)
+      return result;
+    result = SetValue(rawHandle, success, pData, dataLen);
+    PutEntry(rawHandle);
+    return result;
+  }
+
+};
+
+// =====================================================================================================================
+// A slight alternative using more C++ RAII safety:
+// - make all methods of ICache other than GetEntry protected
+// - use this class for the EntryHandle, providing more type safety and using ~EntryHandle
+//   to ensure that PutEntry gets called etc.
+class EntryHandle
+{
+public:
+  EntryHandle() = default;
+  ~EntryHandle() {
+    Put();
+  }
+  EntryHandle(const EntryHandle&) = delete;
+  EntryHandle& operator=(const EntryHandle&) = delete;
+
+  EntryHandle(EntryHandle&& rhs)
+  : m_pCache(rhs.m_pCache), m_rawHandle(rhs.m_rawHandle), m_mustPopulate(rhs.m_mustPopulate) {
+    rhs.m_pCache = nullptr;
+    rhs.m_rawHandle = nullptr;
+    rhs.m_mustPopulate = false;
+  }
+
+  EntryHandle& operator = (EntryHandle&& rhs) {
+    if (this != &rhs) {
+      m_pCache = rhs.m_pCache;
+      m_rawHandle = rhs.m_rawHandle;
+      m_mustPopulate = rhs.m_mustPopulate;
+
+      rhs.m_pCache = nullptr;
+      rhs.m_rawHandle = nullptr;
+      rhs.m_mustPopulate = false;
+    }
+
+    return *this;
+  }
+
+  void Put() {
+    if (m_pCache) {
+      if (m_mustPopulate)
+        m_pCache->SetValue(m_rawHandle, false, nullptr, 0);
+      m_pCache->PutEntry(m_rawHandle);
+    }
+    m_pCache = nullptr;
+    m_rawHandle = nullptr;
+    m_mustPopulate = false;
+  }
+
+  // semantics of these methods are largely analogous to the above in ICache,
+  // their implementation simply forwards to the m_pCache.
+  Result WaitForEntry() const {
+    return m_pCache->WaitForEntry(m_rawHandle);
+  }
+
+  Result GetValue(void *pData, size_t*pDataLen) const {
+    return m_pCache->GetValue(m_rawHandle, pData, pDataLen);
+  }
+
+  Result GetValueZeroCopy(const void **ppData, size_t *pDataLen) const {
+    return m_pCache->GetValueZeroCopy(m_rawHandle, ppData, pDataLen);
+  }
+
+  Result SetValue(bool success, const void *pData, size_t dataLen) {
+    assert(m_mustPopulate);
+    m_mustPopulate = false;
+    return m_pCache->SetValue(m_rawHandle, success, pData, dataLen);
+  }
+
+  /// \brief Construct from a raw handle.
+  ///
+  /// Only called from ICache::MakeHandle
+  EntryHandle(ICache *pCache, void *rawHandle, bool mustPopulate) {
```

Comment:
can you explain what the deference of static creation function and public constructor?

Reply:
To me it seems like you are trying to 'hide' this constructor and prevent people from using it directly. The comment says:
```
  /// Only called from ICache::MakeHandle
```

So what I'm suggesting is that if you do want to hide it (and don't want to use `friend`), you can make it a static function with an appropriate name so that it's obvious you are not supposed to use it directly.

Reply:
At begin, we indeed make it static as comment. During internal review, our internal reviewer suggests just using a public constructor. 
I don't have strong opinion on that, both are ok to me, if you think static is better, I can change back to static way. Here is final review.

Reply:
I'd either move the public constructor to the top of the class or keep it at the bottom as a static function.